### ISSUE TITLE
MacOS support

### ISSUE THREAD

**Issue by smith** [plengauer/opentelemetry-bash#550]
The release files are debs. If I download the source and run `otel.sh` I get

```
exec: Failed to execute process './usr/bin/otel.sh': The file specified the interpreter '/bin/false', which is not an executable command.
```

So, I'm assuming the scripts are making some Linux-specific assumptions.

A version that runs on MacOS would be great, if you're not planning that, a note about OS support in the README would be great.

If making the scripts MacOS compatible seems like it's not impossible, any notes on how to get it working would be helpful.

**Comment by plengauer** [comment]
@smith it looks like you have tried to run it directly, but it needs to be sourced rather than executed. in the readme, there is a space between the `.` and  the `otel.sh`. the file is not intended to run directly, therefore the `/bin/false` shebang.

that being said, i never tried to run it anything non-linux. the code itself doesnt have a lot (if any) linux specifics, but it uses a bunch of external dependencies (like `cut`, `tr`, `grep`, `awk`, `sed`, ...) that need to be installed (as per the debian control file at https://github.com/plengauer/opentelemetry-bash/blob/e7ff252a6026fecfc149b8a5472a4512f286e9b2/src/DEBIAN/control).
assuming the same external executables can be installed on a mac system, i dont see why it wouldnt work.

let me think about that more

PS: your point about adding a note about OS support is definitely a good one!

**Comment by smith** [comment]
(some things in this output might not actually be problems, as I mucked around with the paths so I didn't have to install to / on my Mac, and I haven't set any OTEL environment variables to actually try to see if any data is sent.)

Any use of /proc isn't going to work but I'm pretty sure there's native MacOS alternatives to anything being used here.

I think there's also some MacOS differences in naming and CLI flags (Some BSD-isms, some Mac specific) like it's `nc` not `netcat`.

```bash
bash-5.2$ . ./src/usr/bin/otel.sh
bash: /proc/23404/cmdline: No such file or directory
bash: /proc//cmdline: No such file or directory
WARNING When using OpenTelemetry in an interactive shell for the first time after startup, it may take some time to create the instrumentation cache! Subsequent interactive shells will start faster. This performance impact does not apply in non-interactive shells, like scripts or invocations with -c.
bash: type: busybox: not found
bash: type: podman: not found
bash: type: taskset: not found
bash: type: ionice: not found
bash: type: flock: not found
bash: type: dumb-init: not found
bash: type: ncat: not found
bash: type: netcat: not found
bash: type: node: not found
bash: type: nodejs: not found
bash: type: parallel: not found
bash: type: parallel.moreutils: not found
bash: /proc/29875/cmdline: No such file or directory
bash: /proc/29875/cmdline: No such file or directory
bash: type: ash: not found
bash: /proc/29875/cmdline: No such file or directory
bash: /proc/29875/cmdline: No such file or directory
bash: /proc/29875/cmdline: No such file or directory
Detector <__main__.AwsEC2ResourceDetector object at 0x101115fa0> took longer than 5 seconds, skipping
Detector <__main__.OracleResourceDetector object at 0x1035f8620> took longer than 5 seconds, skipping
^C
bash-5.2$ SDK Error: SPAN_START /var/folders/hw/917m5r6d33780w6mvptjr50c0000gn/T/tmp.tuuxSftRe7_opentelemetry_shell_29875.span_handle.pipe   SERVER
Traceback (most recent call last):
  File "/Users/smith/Desktop/opentelemetry-bash-4.23.1/src/usr/bin/otelsdk.py", line 122, in main
    handle(scope, version, tokens[0], tokens[1] if len(tokens) > 1 else None)
  File "/Users/smith/Desktop/opentelemetry-bash-4.23.1/src/usr/bin/otelsdk.py", line 232, in handle
    name = tokens[4]
           ~~~~~~^^^
IndexError: list index out of range
```

**Comment by plengauer** [comment]
Needs
*) a dedicated build step to build whatever thr native mac package type is in build.yml
*) package meta info in meta dir
*) a mac job like the linux/windows shell jobs in test
Test_shell.yml
*) replacing all accesses to /proc with something equivalent. This should not be dynamic, but rather functions should be defined differently once. For that, check if the dir /proc exists
*) everywhere where there is something os specific, there needs to a new case. For example, where dpkg is used
*) tests that are designed for Linux envs, like debconf should be skipped on mac
*) readme needs an entry that mac is supported additional to windows and linux
Linux
*) the root level install scripts needs adjustments to handle

**Comment by plengauer** [comment]
# macOS support — consolidated state before restart

This consolidates **everything** found, implemented, reviewed, rejected and still missing across the macOS work, so that #3174 and #3942–#3951 can be closed and their branches deleted without losing anything. Sources: this issue, #3952, all 11 open macOS PRs (descriptions, every commit, every diff to CI and production code, every review, review thread and comment), the CI runs referenced in them, and a read of current `main` (`e7dd341`).

Legend: ✅ implemented and worth carrying over · ⚠️ implemented but needs rework · ❌ dead end / broken (reason given) · ❓ unverified — must be checked on a real macOS runner · 🔎 not raised in any PR or review; found while writing this summary (my own analysis, called out as such).

> **Code stays retrievable after branch deletion.** GitHub keeps `refs/pull/<n>/head` for closed PRs, so `git fetch origin pull/3174/head` etc. still works. The March 2026 iteration of #3174 was force-pushed away and only survives as unreferenced commits (`b2ba2a07`, `03c0470a`, `cec22bfa`, `6bee3d72`, `c462d723`, `990c5cdc`). They were still fetchable by SHA today, but GitHub may garbage-collect them, so everything useful in them is listed below.

---

## 0. TL;DR

- **Nothing on any branch has ever passed the macOS test suite.** The most recent `ci / shell / smoke (macos-latest)` on #3174 (head `3220a0c`, [job 101712024483](https://github.com/plengauer/Thoth/actions/runs/33049435679/job/101712024483), 2026-09-07) is the first run that gets past installation. It then fails on the first real test: `unit/test_unit_escape_arg.sh: line 2: /usr/bin/opentelemetry_shell_api.sh: No such file or directory`. (The only test that "succeeded" before that, `test_unit_busybox_parses_otelapi.sh`, exits 0 when busybox is absent, so it proves nothing.)
- **CI never ran on #3942–#3951.** Their base branch is `claude/add-macos-support`, so only `greeting`/`export`/Copilot review ran. Several commits say `NOTE: untested` outright.
- **There are two diverging lines that conflict with each other:**
  - #3174 head, which Copilot rebased onto `main` on 2026-08-26/27 (`84f396fa`…`3220a0c1`).
  - The moflwi stack #3942–#3951, based on the June state of #3174 (`d0d4253`). The two lines fixed the same problems (the SIP-sealed `/usr`, `merge-http` nesting, the formula) in **different, incompatible** ways.
- **Maintainer direction to follow on restart:**
  - Issue comment: OS-specific functions are defined *once*, based on `[ -d /proc ]`, never checked dynamically.
  - Reviews on #3942–#3948: rebase, don't overlap, and replace runtime path resolution with something like **rewriting paths when building the tarball**.
  - #3952: ship a manually installable tarball from this repo, plus a **separate Homebrew tap repo** that consumes the tarball and releases it properly.
- **Do NOT carry over from #3174's head:**
  1. The literal `Authorization: ******` token regression.
  2. The `/usr/local/usr/...` install layout, where the runtime can never find `api.sh` or the venv.
  3. The Darwin branches in `actions/instrument/*`, which are dead code (§4, D19).

---

## 1. Goals and requirements

### 1.1 Original request (@smith, 2024-08)

- The release assets are only `.deb`s. Running `./usr/bin/otel.sh` fails because of the `/bin/false` shebang: the file must be sourced (`. otel.sh`).
- Asked for either macOS support or an OS-support note in the README.
- Data point from @smith's 2024 attempt (sourced from source, not installed):
  - Errors: `/proc/<pid>/cmdline: No such file or directory` and `/proc//cmdline` (empty `$PPID` path).
  - Noisy `bash: type: X: not found` for busybox, podman, taskset, ionice, flock, dumb-init, ncat, netcat, node, nodejs, parallel, parallel.moreutils, ash.
  - The AWS EC2 and Oracle resource detectors each "took longer than 5 seconds, skipping".
  - SDK crash: `SDK Error: SPAN_START /var/folders/.../T/tmp.XXXX_opentelemetry_shell_<pid>.span_handle.pipe SERVER` → `IndexError` in `otelsdk.py` handle() (`name = tokens[4]`). That means the span name arrived empty.
  - The same output shows macOS `mktemp -u` without a template works, producing `/var/folders/.../T/tmp.XXXX`. That is relevant to a false-positive review comment (§5.3).
  - He also noted BSD-isms such as `nc` vs `netcat`.

### 1.2 Maintainer checklist (comment of 2026-03-15) and where it stands

| # | Requirement | Status across all PRs |
|---|---|---|
| 1 | Dedicated build step in `build.yml` for the native mac package | ⚠️ `build-brew` builds a tarball (#3174). It is **not published**: `publish.yml` uploads only `*.deb`, `*.rpm`, `*.apk` and attests only those (🔎). |
| 2 | Package meta info in `meta/` | ⚠️ `meta/homebrew/opentelemetry-shell.rb` exists, with `__VERSION__`/`__SHA256__` placeholders. #3942 templates them into a `homebrew-formula.<ref>` artifact. Nothing publishes it; the tap does not exist (#3952). |
| 3 | A mac job in `test_shell.yml` like the linux/windows ones | ⚠️ Added (`macos-latest`, `macos-13`). `macos-13` never schedules. `macos-latest` fails at the first test. |
| 4 | Replace every `/proc` access with an equivalent, **defined once** (check whether `/proc` exists), not dynamic | ⚠️ Partial. Many `/proc` uses remain (§6.1). #3174's curl/wget changes call `$(\uname -s)` at runtime on every invocation, which **violates** the "define once" rule. |
| 5 | Every OS-specific spot (e.g. `dpkg`) gets a new case | ⚠️ Partial (instrumentation cache key, installers). Package-version resolution on macOS was implemented in March, lost in April, never restored (§4, D11). |
| 6 | Skip Linux-only tests (e.g. debconf) on mac | ❌ Not done systematically. Some tests self-skip already (debconf `confmodule` check, docker `type docker`). |
| 7 | README says mac is supported | ⚠️ Done in #3174. It claims macOS is "actively tested", which is not true yet. |
| 8 | Root-level install script handles mac | ⚠️ #3174 adds a `brew tap`/`brew install` branch that cannot succeed (§4, D7). #3946 adds a tarball fallback, but the tarball is not a release asset (D8). |

### 1.3 Later maintainer direction

- **#3174 review (2026-06-16):** "no package of any kind is built … add an appropriate job to build the macos equivalent and install it in the test before running them". → `build-brew` job + test install step.
- **#3952 (2026-08-09):** "make all the changes for mac here, build a tarball that can be installed manually, and have a separate homebrew tap repository that consumes the tarball and releases it properly".
- **Reviews on #3942–#3948 (2026-08-15):** "lets rebase", "lets make sure they are not overlapping", "i suggest finding a new solve for the dynamic path resolution everywhere - like rewriting paths when building the tarball".

---

## 2. Inventory

| PR | Title | Author | Base → head | Notes |
|---|---|---|---|---|
| #3174 | Support MacOS | Claude agent, later Copilot | `main` ← `claude/add-macos-support` @ `3220a0c` | Draft, dirty. Net diff vs `main` is 16 files, +339/−56. |
| #3942 | SIP-safe install paths and shell detection (foundation) | moflwi | `claude/add-macos-support`@`d0d4253` ← `add-macos-support-fixes` @ `0735b49` | 3 commits (`d40a73c`, `75f9b87`, `0735b49`). **Every other moflwi PR (#3943–#3951) builds on these 3 commits** (stacked). |
| #3943 | fix merge-http nesting | moflwi | same base ← `fix/merge-http-arch-collision` @ `c59df6f` | |
| #3944 | node dir via `$_otel_shell_home` | moflwi | same base ← `feat/macos-node-instrumentation` @ `019afa5` | |
| #3945 | Java agents resolve own install path | moflwi | same base ← `feat/macos-java-instrumentation` @ `09f35eb` | |
| #3946 | tarball fallback in INSTALL.sh | moflwi | same base ← `feat/macos-install-sh-tarball` @ `4ff8a6c` | |
| #3947 | asymmetric docker bind mounts | moflwi | same base ← `fix/macos-docker-mount-paths` @ `66ed1cb` | |
| #3948 | postinst prefix-aware | moflwi | same base ← `feat/macos-postinst-prefix` @ `92cbffd` | 2 own commits (`921271c`, `92cbffd`) |
| #3949 | run the real postinst from tarball + bundle node_modules | moflwi | `feat/macos-postinst-prefix` ← `feat/macos-postinst-wiring` @ `be37c1c` | Own commits `e0846b5` (run the real postinst) and `be37c1c` (bundle node_modules); also merges #3946 (`94347bb`). Draft. |
| #3950 | consolidate portable find | moflwi | `fix/macos-docker-mount-paths` ← `cleanup/consolidate-portable-find` @ `8d14fd2` | Draft |
| #3951 | bundle Python site-packages | moflwi | `feat/macos-postinst-wiring` ← `feat/macos-python-instrumentation` @ `f138677` | Draft |

Related, keep open or handle separately:

- **#3952** — Homebrew tap decision. Open, not closed by any PR.
- Merged CI coverage for the *GitHub Actions* instrumentation on macOS runners: #3550 (+ backport #3559) adds `macos-15`/`macos-latest` and `windows-2022`/`windows-latest`; #3894 adds `macos-26`, a shell-info step and a `zsh` step to `job-macos`; #3887 was a WIP superseded by #3894.
  - 🔎 These `job-macos` jobs only prove the action doesn't break a macOS job: the job action is a no-op on non-Linux (§4, D19).
- 🔎 Overlapping in-flight PRs to coordinate with on restart:
  - #3268 restructures OS detection in `actions/instrument/shared/install.sh` (rpm/apk runners).
  - #4139 rewrites `meta/debian/postinst` Python install to be offline-first from shipped wheels (the same file #3948/#3949/#3951 change).
  - #4202 drops the experimental job whose failure was noted on #3174 (item 3 of the 2026-08-27 pass). That points to the failure being unrelated to macOS.

History of #3174:

1. March: initial implementation (`03c0470a`…`c462d723`) and the maintainer's line comments, fixed in `990c5cdc`.
2. April: Claude rebase into a single commit `af80e225`. It silently dropped several March changes (D11).
3. April–June: `main` merged in 3 times, then `2bbafcc1` ("skip non-linux HTTP binaries when building deb/rpm") and `d0d4253` (`build-brew` + test install).
4. 2026-08-26/27: Copilot re-applied everything onto current `main` (`84f396fa`, `1ab9cd2c`, `5074c469`, `2a1b9fb1`), merged the old line back in (`049fb468`), then `3220a0c1` (gnubin loop fix).

---

## 3. What was implemented (carry-over candidates), by area

### 3.1 Build (`.github/workflows/build.yml`)

- ✅ Rename `build-http` → `build-http-linux`, artifacts `http_<arch>_<ref>` → `http_linux_<arch>_<ref>`, and update all references (maintainer's request on #3174).
- ✅ New job `build-http-macos` on `macos-latest`, matrix `arm64`, `x86_64`: `clang -arch <arch> -dynamiclib -fPIC -o libinjecthttpheader.dylib agent.injection.http_header.c`, uploaded as `http_macos_<arch>_<ref>`. The first version used `-shared`, changed per review (D15). It compiles: both matrix jobs are green on `3220a0c`.
  - 🔎 It uses tag-pinned actions (`actions/checkout@v6.0.2`, `upload-artifact@v7.0.0`, `opentelemetry-github/...@v5.47.2`), while the rest of `build.yml` is SHA-pinned with version comments. Fix for Renovate consistency.
- ✅ `merge-http` needs both build jobs and parses `http_<platform>_<arch>_<ref>`. It **must merge per arch file-by-file**: Linux `x86_64` and macOS `x86_64` collide, and `mv "$directory" bin/"$arch"` nested the second one (D16). Two equivalent fixes exist:
  - #3174: `mkdir -p bin/"$arch"; mv "$directory"/* bin/"$arch"/; rmdir "$directory"`
  - #3943: `mkdir -p bin/"$arch"; cp -r "$directory"/. bin/"$arch"/`
  - Also, #3174 swapped `find -printf` for `ls | grep '^http_'`. That was unnecessary, since the job runs on Linux.
- ✅ deb/rpm packaging skips arch dirs without a `.so` (`2bbafcc1`/`1ab9cd2c`): `find "$replaced_binaries"/"$arch" -maxdepth 1 -name '*.so' | grep -q . || continue`, and the same guarded by `[ "$arch" = noarch ] ||` in rpm. Otherwise the macOS-only arch dirs break `cp *.so`.
- ✅ Tarball job:
  - March `build-homebrew` ran on `macos-latest` and bundled node_modules, python site-packages, java agents and the macOS dylibs into `arch/` subdirs. It also deleted `Dockerfile`s, `*.c`, `*.java`, `manifest.txt`, `pom.xml`, and fed `merge-packages`.
  - June `build-brew` (`d0d4253`/`5074c469`) runs on `ubuntu-slim`, `needs: [verify-scripts, merge-http]`. It deletes `*.c`/`*.java`/`manifest.txt`/`pom.xml`, downloads `http_<ref>`, deletes `*.so` and empty dirs, runs `tar -czf opentelemetry-shell_<ver>.tar.gz -C src .`, and uploads `packages.brew.<ref>`. It **dropped** the node, python and java bundling.
  - Later additions to `build-brew` from the moflwi stack:
    - #3945: `needs build-java-agents`, download `java_<ref>`.
    - #3949: `needs merge-node-modules`, download `node_modules_<ref>`; ship `meta/debian/postinst` as a top-level `postinst` file in the tarball.
    - #3951: `needs merge-python-site-packages`, download `python_site_packages_<ref>`.
    - #3942: template the formula (`sed "s/__VERSION__/$version/g; s/__SHA256__/$sha256/g"`) and upload `homebrew-formula.<ref>`.
  - Note: the tarball members are `./usr/...`, `./opt/...` (and `./postinst`). This matters for `--strip-components` (§5.3).
- ✅ Java agents (#3945), which also improves Linux:
  - `Boot-Class-Path` becomes relative: `./subprocessinjectionagent.jar ./byte-buddy.jar` and `./gradlehttppropagationagent.jar ./byte-buddy.jar`.
  - `premain` resolves its own jar via `X.class.getProtectionDomain().getCodeSource().getLocation().toURI()` instead of a hardcoded `/usr/share/...` path.
  - The author verified this locally on OpenJDK 17 from two prefixes. The full compile against byte-buddy/otel-javaagent was not verified: CI never ran.
- ✅ Node modules are portable: the author ran `npm install` of all 446 packages and found no `.node` addons and no `binding.gyp`. So the Linux-built `node_modules.tar.xz` works on macOS unchanged (#3949).
- ✅ Python site-packages are portable. I verified this in `build.yml`: `build-python-site-packages` uninstalls every package whose `RECORD` lists a `.so` (grpcio, wrapt, …) and then asserts no `.so` remains. `postinst` re-installs native deps per Python version with `pip3 install -r requirements.txt`, so they come out platform-correct (#3951).

### 3.2 Packaging, release, Homebrew

- ⚠️ `meta/homebrew/opentelemetry-shell.rb`, latest state in #3174 `2a1b9fb1`:
  - `url .../releases/download/v__VERSION__/opentelemetry-shell___VERSION__.tar.gz`, plus `sha256`/`version` placeholders and `license "Apache-2.0"`.
  - `depends_on`: coreutils, findutils, grep, gnu-sed, gawk, jq, `python@3.12` (hard dependency), `xxd => :recommended`.
  - `install`: `prefix.install Dir["*"]`; move the current arch's dylib (`Hardware::CPU.arm? ? "arm64" : "x86_64"`) to the flat `agent.instrumentation.http/libinjecthttpheader.dylib`; `rm_rf` both arch dirs; `bin.install_symlink` `otel.sh` and `otelapi.sh`.
  - `post_install`: create the venv with `Formula["python@3.12"].opt_bin/"python3.12" -m venv`, then use the venv's `pip3`.
  - `caveats`: source `. #{bin}/otel.sh`, and prepend the gnubin dirs to `PATH` for both prefixes.
  - `test do`: `system "/bin/sh", "-c", ". #{bin}/otel.sh"`.
  - #3942 changed the same file differently: `python@3 => :recommended` + `system "python3"`, and a caveat that the scripts look under `/usr/local/share/opentelemetry_shell`, so Apple Silicon (`/opt/homebrew`) needs a symlink. Copilot flagged the `:recommended` Python and the `system "python3"` call on four PRs. **#3174's final formula is the better base.**
  - Still missing (open items):
    - The formula lives at `meta/homebrew/`, not in a tap repo.
    - Nothing publishes it on release.
    - The Apple Silicon prefix mismatch has not been decided.
    - Under Homebrew the files land in `#{prefix}/usr/share/...`, which matches none of the hardcoded runtime paths.
    - `post_install` installs the SDK venv under `#{prefix}/opt/...`, while `api.sh` hardcodes `/opt/opentelemetry_shell/venv`. 🔎
- ⚠️ `meta/debian/postinst` made prefix-aware (#3948), so tarball installs can reuse it:
  - `_otel_shell_home` = `$OTEL_SHELL_HOME` (packager override) or `/usr/local/share/opentelemetry_shell` on Darwin, else `/usr/share/opentelemetry_shell`. All 26 literals replaced.
  - `_otel_cp_update` helper: `cp -R` on Darwin (BSD `cp` has no `-u`), `cp -r -u` elsewhere.
  - `/opt/opentelemetry_shell` is unchanged: `/opt` is not sealed on macOS.
  - Python handling went through three iterations (D6); the final one is "same archive on all platforms".
  - Whatever the final layout, the separate helpers (INSTALL.sh, the action installer, CI) must not re-implement postinst logic. #3949's idea of *running the real postinst* from the tarball is the keeper.
  - The author claims a security review found no attacker-writable window in the `mktemp` → `sudo sh "$tmp" configure` pattern (#3949 test plan). Not independently verified.
- ❌ Release publishing: the tarball, the formula, attestation. Nothing does it (§6.4).

### 3.3 Installers and CI install step

- ⚠️ `.github/workflows/test_shell.yml` (#3174):
  - Matrix adds `macos-latest` and `macos-13` (drop `macos-13`, D9).
  - Linux-only steps are gated by `if: runner.os == 'Linux'` (deb download/install, `apt-get remove`). macOS downloads `packages.brew.<ref>`.
  - macOS install runs `brew install jq wget curl bash coreutils findutils grep gnu-sed gawk` and writes every existing gnubin dir to `$GITHUB_PATH` with `|| true` (D17).
  - Extraction layout differs between the lines (D1, D2 vs #3942's split extraction): #3942 does `tar --strip-components=2 -xzf pkg -C /usr/local usr` and `tar --strip-components=1 -xzf pkg -C / opt` (member names written without `./`, see §5.3), then flattens the dylib and creates the venv. #3949 replaces the venv part with `tar -xzf pkg -O postinst > tmp; sudo sh tmp configure`.
  - 🔎 `sudo pip3` without `-H` prints "cache has been disabled" (visible in the log).
- ⚠️ `actions/instrument/shared/install.sh` (#3174):
  - Darwin `ensure_installed` uses `brew install` (via `xargs -n 1`) and installs GNU tools plus gnubin `PATH` (with `|| true`).
  - Install order: `brew tap` + `brew install`, then a tarball fallback from the tagged release, then the latest release. Every asset download selects exactly one URL (`head -n 1 | xargs -r -I '{}'`, no `-0`), fixed per review (D10 covers the token regression).
  - Keep the idea only: the file must be rewritten against the final layout, and all of it is dead code today (D19).
- ⚠️ `actions/instrument/job/inject_and_init.sh` (#3174): skips the apt cache and the cached-deb fast install on Darwin. It is also dead code (D19). `actions/instrument/workflow/main.sh` does the same: `if [ "$(uname -s)" = "Darwin" ] || type dpkg` → install deps.
- ⚠️ `INSTALL.sh`:
  - #3174 adds `elif Darwin && type brew → brew tap … ; brew install …; exit 0`.
  - #3946 changes it: when brew is missing or the tap install fails, set `extension=tar.gz`, select the asset with `endswith("tar.gz")` then `.[0]`, fall back to the direct-download names `opentelemetry-shell_<ver>.tar.gz`, then install with the same split `usr` → `/usr/local`, `opt` → `/` extraction plus postinst (#3949).
  - 🔎 `INSTALL.sh` needs `jq` and `wget`, which stock macOS doesn't ship (it has `curl`).

### 3.4 Runtime (`src/usr/share/opentelemetry_shell/…`)

- ⚠️ Shell detection without `/proc` in `api.sh`:
  - #3174: `if [ -d /proc ]; then <existing>; else _otel_shell="${SHELL##*/}"; …; fi`, plus the commandline-override check without the `/proc/$PPID/cmdline` comparison.
  - #3942 replaces `$SHELL` (the login shell, not the running interpreter) with `ps -p $$ -o comm=` (basename), falling back to `$SHELL`, then `sh`.
  - 🔎 Login shells report `comm` as `-bash`; strip a leading `-`.
  - Must be restructured as "define once at load" per the maintainer.
- ⚠️ `agent.sh` hint logic: `elif [ -f "$0" ] && [ -d /proc ] && [ "$(readlink -f "$0")" != "$(readlink -f /proc/$$/exe)" ]` (#3174). #3942 also guards `_otel_resolve_instrumentation_hint` (the `_otel_hint_is_self` flag).
- ⚠️ Instrumentation cache key in `agent.sh`: an added `elif type brew → ls -la /usr/local/Cellar /opt/homebrew/Cellar 2>/dev/null || true`, alongside the dpkg-lock case (#3174). This is the "new case where dpkg is used" requirement.
- ✅ **BSD `find` has no `-executable`** (#3942 `0735b49`). `_otel_list_path_executables` swallowed the error (`2>/dev/null || true`), so macOS would have instrumented *nothing* silently.
  - Fix: pick the predicate once at load, `-perm +111` on Darwin. #3947 adds `_otel_find_executables` for docker; #3950 consolidates both into `_otel_find_executable_test`.
  - The predicate choice is still ❓ (§5.3).
- ⚠️ curl/wget HTTP header injection (#3174):
  - On Darwin, use `libinjecthttpheader.dylib` via `DYLD_INSERT_LIBRARIES` (save and restore `OLD_DYLD_INSERT_LIBRARIES`) instead of `LD_PRELOAD`.
  - Problems: it calls `uname` at runtime per invocation (violates "define once"), hardcodes `/usr/share/...` (sealed on mac), and the `ldd` check is a no-op on mac (D5).
  - 🔎 It is very likely **functionally ineffective** on macOS (D13).
- ⚠️ Path resolution: #3942/#3944/#3945/#3947/#3951 replace the hardcoded `/usr/share/opentelemetry_shell` with `$_otel_shell_home` (and `/usr/bin` with `$_otel_shell_bin_home` for docker). Covered: `api.sh` (`sdk.py` path, sourcing `api.observe.{logs,pipes,subprocesses}.sh`), `agent.sh` (sourcing `api.sh`, the special instrument file list), curl, wget, `agent.injection.netcat.sh`, node (5), java (4), python (8), docker, gh-aw docker.
  - **Superseded by the maintainer's "rewrite paths at tarball build time" direction**, but the list of touched places is the checklist for the rewrite. The full inventory is in §6.2.
- ✅ Docker bind mounts from a macOS host must be **asymmetric** (#3947): containers are Linux even under Docker Desktop.
  - source = host path (`$_otel_shell_home`, `$_otel_shell_bin_home/<file>`), target = the Linux paths (`/usr/share/opentelemetry_shell`, `/usr/bin/<basename>`).
  - The same applies to the gh-aw compose `yq` edits (`strenv(OTEL_BIND_SOURCE)`/`strenv(OTEL_BIND_TARGET)`).
  - `/opt/opentelemetry_shell` is identical on both sides.
  - This stays necessary under any path strategy.
  - The author notes the mount sources derive only from `uname -s`, not from attacker-influenceable input (#3947 test plan).
- ✅ Package version cache: `_otel_package_version` is used in cache file names and the SDK version string.
  - March had `|| brew list --versions "$1" | cut -d ' ' -f 2` in `_otel_resolve_package_version`. It was lost (D11).
  - A tarball install is known to no package manager.
  - 🔎 Suggestion: bake the version in at tarball build time. That fits the build-time rewrite.

### 3.5 Tests

- ✅ `tests/run_tests.sh` (#3174 final):
  - `if [ -r /etc/os-release ]; then . /etc/os-release; fi`.
  - POSIX `perl() { cat; }` fallback.
  - `timeout`: use it if present, else `gtimeout` (Homebrew coreutils) as a wrapper, and only as a last resort the no-op `timeout() { shift; "$@"; }`. The no-op makes the timeout tests vacuous (D18).
  - History of the perl fallback: the June line (`d0d4253`) dropped the pre-existing `perl` fallback when it added the timeout fallback; #3942 restored it. #3942's version of this file uses bash-only `function perl()`/`function timeout()` (flagged by Copilot 6×). Use #3174's.
- ✅ Busybox guard in `tests/auto/test_auto_injection{,_c,_stdin}.sh`: `if [ -d /proc ] && [ "$(readlink -f /proc/$$/exe | …)" = busybox ]; then exit 0; fi`.
- ⚠️ 60+ test files hardcode `/usr/bin/opentelemetry_shell.sh` and `/usr/bin/opentelemetry_shell_api.sh`. `main` has 63 such files under `tests/`, and this is exactly what the latest macOS run fails on.
  - #3942 parameterised them with `. ${OTEL_TEST_ENTRYPOINT:-/usr/bin/opentelemetry_shell.sh}` / `OTEL_TEST_API_ENTRYPOINT`, set on Darwin in `run_tests.sh` to `/usr/local/bin/...`. Review asked to quote the expansion.
  - Alternative (🔎): make the tarball install put the entrypoints on `PATH` and have tests source `opentelemetry_shell.sh` by name. The `src/usr/bin/*` entries are **relative symlinks** (`otel.sh -> ../share/opentelemetry_shell/agent.sh`), so they are relocatable.

### 3.6 README

- ⚠️ #3174 adds macOS to the supported list and a macOS install section (`brew tap …`, `brew install …`, plus two `export PATH=…gnubin…` lines, one per prefix), and removes "Mac-based operating systems are currently not supported".
- Rewrite it once the real install path exists. Don't claim "actively tested" until CI is green.

---

## 4. Dead ends, broken approaches and regressions (with the reason)

| ID | What | Why it was dropped or why it's broken |
|---|---|---|
| D1 | Extract the tarball into `/` on macOS (`d0d4253`, June) | macOS's sealed system volume: `Can't create 'usr/bin/otel.sh': Read-only file system` ([job](https://github.com/plengauer/Thoth/actions/runs/27628357041/job/81730960324)), even as root. |
| D2 | Extract the whole tarball into `/usr/local` (March + #3174 head) | Produces `/usr/local/usr/share/opentelemetry_shell` and `/usr/local/opt/opentelemetry_shell`, but the runtime hardcodes `/usr/share/opentelemetry_shell/api.sh` (sourced by `agent.sh`) and `/opt/opentelemetry_shell/venv/bin/python` (`api.sh`). `otel.sh` resolves, but `api.sh` and the SDK venv are never found (🔎 verified by grep on `3220a0c`). |
| D3 | Runtime path variable `_otel_shell_home` hardcoded per OS (#3942–#3951) | The maintainer asked for a different solution (rewrite paths at tarball build time). Copilot flagged it repeatedly (7 threads across 5 PRs) as breaking any other prefix, e.g. Homebrew's `/opt/homebrew`. Adding more `uname` calls at load also cuts against "define once". |
| D4 | `$SHELL` as the shell when `/proc` is missing (#3174) | `$SHELL` is the login shell, not the interpreter that sourced the file. Replaced by `ps -p $$ -o comm=` in #3942. |
| D5 | Dependency check of the injection library on Darwin | `ldd` doesn't exist → the check silently passes (the maintainer resolved this as harmless). #3942's `otool -L … \| grep 'not found'` never matches either, because `otool -L` lists install names and doesn't resolve them (Copilot, 3 PRs). Either drop the check on Darwin or do a real probe, e.g. `DYLD_INSERT_LIBRARIES=$lib /usr/bin/true`, where 🔎 SIP may strip the variable for `/usr/bin/true` (see D13). |
| D6 | Deep-Python handling on macOS in postinst | (a) Skip it on Darwin (`921271c`) → replaced by (b) rebuild from scratch via pip with empty version dirs (`92cbffd`) → replaced by (c) use the same pure-Python archive as Linux (`f138677`), because the build already strips every `.so` package (verified). (a) and (b) are dead. |
| D7 | `brew tap plengauer/opentelemetry-shell https://github.com/plengauer/Thoth` + `brew install` (INSTALL.sh, action installer, README) | 🔎 This can never work as written. A tap needs formulae in `Formula/`, `HomebrewFormula/` or the repo root; here it's at `meta/homebrew/` with unfilled placeholders. #3952 confirms there's no tap, so the path always falls through. Per the maintainer, use a separate tap repo. |
| D8 | Tarball download from GitHub releases (INSTALL.sh fallback, action installer fallbacks, formula `url`) | 🔎 `publish.yml` releases only `*.deb`, `*.rpm`, `*.apk` (`fail_on_unmatched_files`), so no `.tar.gz` asset exists and every one of these URLs 404s today. |
| D9 | `macos-13` in the test matrix | It never gets a runner: queued >24 h, [job](https://github.com/plengauer/Thoth/actions/runs/32941063404/job/98112292081). Later runs were cancelled after ~8 h. That blocks `needs: smoke`. Dropped in #3942; still present in #3174. |
| D10 | **Copilot's 2026-08-26 rebase (`84f396fa`) replaced `Authorization: Bearer $INPUT_GITHUB_TOKEN` with the literal `Authorization: ******`** — 6 occurrences on 4 lines of `actions/instrument/shared/install.sh`, **including both Linux deb download paths** | 🔎 Evidently log-redacted text was pasted back into source. Every authenticated release download sends a bogus header. CI doesn't catch it because CI installs the local `.deb`. Must never be carried over. |
| D11 | April rebase `af80e225` ("Add comprehensive macOS support") | 🔎 Its commit message still claims these, but it silently dropped four March changes: (1) the Darwin skip of strace subprocess observation in `api.observe.subprocesses.sh`; (2) the `ps -p $$ -o comm=` fallback in `_otel_resource_attributes_process` (still `readlink -f /proc/$$/exe` on `main`); (3) the `ps -p $$ -o args=` fallback in `_otel_resolve_command_self` (still reads `/proc/$$/cmdline`); (4) the `brew list --versions` fallback in `_otel_resolve_package_version`. It also dropped the node/python/java bundling in the March `build-homebrew` job (later re-added by #3945/#3949/#3951). (1) is redundant anyway: `api.sh` and `agent.instrumentation.deep.sh` already require `type strace`. (2)–(4) must be redone. |
| D12 | Copilot suggestion to use `-perm -111` instead of `+111` (#3947) | Semantically wrong: `-111` requires **all** execute bits, so it would miss e.g. `0744`. |
| D13 | Plain `DYLD_INSERT_LIBRARIES` with same-named symbols (`write`, `send`, `SSL_write`, `gnutls_record_send*`, `nghttp2_submit_request` + `dlsym(RTLD_NEXT)`) | 🔎 Unverified, but high confidence. macOS uses a two-level namespace, so defining a same-named symbol in an inserted dylib does **not** interpose calls from libcurl/libSystem. That needs `__DATA,__interpose` (`DYLD_INTERPOSE`) tuples or `DYLD_FORCE_FLAT_NAMESPACE=1`. Also, SIP purges `DYLD_*` for platform binaries: `/usr/bin/curl`, `/bin/sh`, `/bin/bash`, `/usr/bin/env`. Homebrew `curl` is keg-only, so `curl` on `PATH` is usually `/usr/bin/curl`. System curl uses LibreSSL/SecureTransport, not OpenSSL/GnuTLS. No test ever exercised this. |
| D14 | Formula `test do system "#{bin}/otel.sh", "--version"` | Can never pass: `/bin/false` shebang, the file must be sourced. → `system "/bin/sh", "-c", ". #{bin}/otel.sh"`. |
| D15 | `clang -shared` for the dylib | Not the macOS way (Copilot) → `-dynamiclib`. |
| D16 | `mv "$directory" bin/"$architecture"` in `merge-http` | Linux `x86_64` and macOS `x86_64` collide, and the second mv nests the dir (seen in CI: `x86_64/http_macos_x86_64_<sha>/libinjecthttpheader.dylib`). → per-file merge. 🔎 A universal binary (`clang -arch arm64 -arch x86_64`) would remove the macOS side of the collision and all the per-arch flattening in the formula, CI and installers. |
| D17 | gnubin loop `[ -d "$d" ] && echo … >> $GITHUB_PATH` as the last command | On arm64 the final `/usr/local/opt/...` iteration is false, so the step exits 1 under `bash -e` with no message ([job](https://github.com/plengauer/Thoth/actions/runs/32941063404/job/98112292061)). The same latent bug existed in `install.sh` (`sh -e`). → `|| true` (`3220a0c`). |
| D18 | `timeout() { shift; "$@"; }` as the only fallback | Makes the timeout-based tests vacuous (e.g. `test_auto_injection_timeout.sh`). → prefer `gtimeout`. |
| D19 | Darwin branches in `actions/instrument/job/inject_and_init.sh`, `actions/instrument/workflow/main.sh`, `actions/instrument/shared/install.sh` | 🔎 **Dead code.** `inject_and_init.js`, `shutdown.js` and `workflow/main.js` all start with `if (process.platform != 'linux') return;`, so none of the shell code runs on macOS runners. That is also why the existing `job-macos` tests pass. Making the actions work on macOS runners is its own piece of work: lift the gate, then cache, install and the otelcol binary. |
| D20 | Pre-existing Copilot claim in #3174 (2026-03) that "this is how Homebrew works" for tarball-only + in-repo formula | Superseded: see D7/D8 and #3952. |
| D21 | Old line (`d0d4253`) conflict resolution in `install.sh` | It reverted `main`'s arch-specific deb selection: `endswith(".deb")` instead of `endswith("_<arch>.deb")`, `package.deb`/`_all.deb` names, and dropped `head -n 1`. Combined with `xargs -0`, `wget` requested one URL with `%0A` separators → 404 → a non-package `.deb` → `workflow-smoke`, `checksuite-smoke` and `repository-smoke` failed (maintainer pass 2026-08-26, item 1). Fixed on the new line; the moflwi stack still has the broken version. |

---

## 5. Review findings ledger (all reviews, deduplicated)

### 5.1 Addressed on #3174's head (keep the fix)

- Maintainer, 2026-03-15: rename to `http_linux_*` / `build-http-linux`; no double backslash (`\\uname` → `\uname`) in curl, wget and `api.observe.subprocesses.sh`. Answered as questions: "is this how Homebrew works?" and "should be brew install, not just a tarball".
- Copilot, 2026-06-16 (18 comments), all fixed in `84f396fa`/`2a1b9fb1`:
  - `-dynamiclib`; `merge-http` per-file.
  - Timeout fallback (`command -v`, `gtimeout`).
  - gnubin in CI, in the installer, in the formula caveats and in the README.
  - `python@3.9 :recommended` → `python@3.12`.
  - Venv chicken-and-egg in `post_install`; dylib flattening in the formula; the formula test.
  - `xargs -0` + multiple URLs at all 4 download sites.
  - Dylib flattening after tarball extraction.
  - Some of those threads are still unresolved or outdated in the UI (`clang -shared`, `merge-http`, `run_tests.sh` timeout, README gnubin note).
- Maintainer pass, 2026-08-26 (6 items): all addressed by Copilot. The item-1 fix introduced D10, and the item-2 fix created D2.
- Maintainer pass, 2026-08-27, item 1 (gnubin loop under `set -e`): fixed in `3220a0c`.

### 5.2 Still open / valid

- `macos-13` never schedules (2026-08-27, item 2): **not fixed on #3174**; fixed in #3942.
- Latest CI (2026-09-07): tests source the hardcoded `/usr/bin/opentelemetry_shell_api.sh` → fails on mac (§3.5).
- The hardcoded `/usr/local/share/...` prefix on Darwin breaks other prefixes (#3942, #3943, #3944, #3945, #3948) → resolve through the build-time path rewrite.
- `agent.sh` sources `api.sh` from a fixed Darwin path (#3945, #3948) → same.
- The `otool -L` check is ineffective (#3944, #3945, #3946) → D5.
- Formula `python@3 => :recommended` + `system "python3"` (#3944, #3945, #3946, #3948) → only in #3942's formula; #3174's formula is fine.
- `function perl()` / `function timeout()` non-POSIX in #3942's `run_tests.sh` (6×) → use #3174's.
- Unquoted `. ${OTEL_TEST_ENTRYPOINT:-…}` (5×) → quote it if that approach is kept.
- Docker share-dir bind mount lacks `readonly` (#3947). It's pre-existing on Linux too; worth hardening.
- `uname -s` forked 3× at `api.sh` load (#3950) → cache it once. That's in line with "define once".
- `grep -vE '^'"$_otel_shell_home"'/…'` uses an unescaped path in a regex (#3951). Low risk.
- Temp-file cleanup via `trap` in the installers (#3949). Nit.
- `opentelemetry-shell_*.tar.gz` glob could match several files in CI (#3949). Low risk.
- Maintainer on #3943 resolved "PR description mismatch" as a nit.
- `type dpkg` unredirected in `workflow/main.sh`: the maintainer resolved it as pre-existing noise.
- The Workflow Example `build` failure seen during the 2026-08-27 pass is a deliberate rate-limit fixture, re-run, not a finding.
- On #3952, a `@copilot input pls` request never got an answer. The maintainer's reply there is the direction quoted in §1.3.
- `ldd` missing on Darwin in #3174: the maintainer resolved it as harmless (permissive degrade), see D5.

### 5.3 False positives or unverified claims (don't re-chase)

- ❌ False positive: "`tar --strip-components=1 -C / opt` lands at `/opentelemetry_shell`" (#3947, 2×). The tarball members are `./opt/...`; stripping 1 component removes `.`, so files land at `/opt/opentelemetry_shell`. I verified this with GNU tar.
  - 🔎 **Real, related fragility:** with GNU tar the member name `usr`/`opt` (without `./`) fails with "Not found in archive"; only `./usr`/`./opt` works. macOS uses bsdtar, whose pattern matching reportedly tolerates the `./` (❓). Use explicit `./usr`, `./opt`, or build the tarball without the `./` prefix.
- ❌ False positive: "`sh -c '. ${OTEL_TEST_ENTRYPOINT:-…}'` won't expand" (#3946, #3948). The inner shell expands it at runtime, and `run_tests.sh` exports the variable.
- ❌ False positive: "INSTALL.sh `tar.gz) jq '.[0]'` may pick a non-tarball asset" (#3946). The pipeline already filters on `endswith("tar.gz")` first.
- ❌ False positive: "`mktemp` without a template fails on macOS" (#3949, 3×). The issue's own 2024 macOS output shows `mktemp -u` producing `/var/folders/.../T/tmp.XXXX`.
- ❌ Moot: "`_otel_find_executable_test` may be unset, or injected via env" (#3950). `api.sh` always sets it unconditionally before `agent.sh` uses it.
  - The author's own note on #3950: a zsh test gave a false negative because zsh does not word-split unquoted variables. 🔎 Any "predicate in a variable" design breaks if the file is ever sourced by zsh, so prefer a function per OS.
- ❓ `-perm +111` "deprecated on BSD find". Copilot gave conflicting fixes: `/111` (GNU syntax), `-111` (wrong, D12), OR-ing `-100/-010/-001`, or listing then filtering with `[ -x ]`. The last is portable and matches `-executable` semantics (the current user can execute). Decide and test on the runner.
- ❓ "`readlink -f` unavailable on macOS" (#3944–#3948). It's moot if gnubin coreutils are first on `PATH` (`\readlink` still resolves through `PATH`). Recent macOS `readlink` reportedly supports `-f`. Verify.
- ❓ "`xargs -r` is GNU-only; breaks postinst on Darwin" (#3948). It's moot with gnubin findutils on `PATH`. **But** 🔎 postinst is run via `sudo sh …`, and `sudo` may reset `PATH` (`env_reset`/`secure_path`), dropping gnubin. Verify BSD `xargs -r` and `sudo` `PATH` behavior on the runner.
- ❓ The claim in #3949's test plan that `tar -xzf pkg -O postinst` works with macOS bsdtar with and without `./` was verified locally by the author on a fixture, not in CI.

---

## 6. Gaps not addressed by any PR (🔎 own analysis against `main` @ `e7dd341`)

### 6.1 Remaining `/proc` accesses (maintainer requirement: define once, based on `[ -d /proc ]`)

- `api.sh:39,42`: shell detection and the commandline-override signature. #3174/#3942 guarded these.
- `api.sh:109` (`_otel_resource_attributes_process`, `readlink -f /proc/$$/exe`) and `api.sh:154` (`_otel_resolve_command_self`, `/proc/$$/cmdline`). Both regressed via D11.
- `agent.sh:45` (hint, guarded in #3174), `agent.sh:183` (`_otel_resolve_instrumentation_hint`, guarded in #3942), and three more that are **unguarded on every branch**:
  - `agent.sh:407`, exec re-invocation via `xargs -0 … </proc/$$/cmdline`
  - `agent.sh:511`, sshd detection via `/proc/$PPID/cmdline`
  - `agent.sh:517`, CGI/python3 parent detection via `/proc/$PPID/cmdline`
- `api.observe.pipes.sh:52`: `readlink -f /proc/self/fd/0`. `:88/:94` already have a `[ -d /proc ]` guard.
- `agent.instrumentation.github.actions.codeql.sh:12`: `xargs -0 … </proc/$$/cmdline`.
- `actions/instrument/job/decorate_action.sh:2` (`/proc/$PPID/cmdline` = `Runner.Worker`), `inject_and_init.sh:18` (`/proc/1/cgroup`), `inject_and_init.sh:351` (`readlink -f /proc/*/exe`): only relevant once D19 is lifted.
- On macOS, `ps -p <pid> -o args=` / `-o comm=` is the usual substitute; for the `/proc/$$/cmdline` → `xargs -0` exec path there is no NUL-separated argv equivalent, so that needs a design decision.

### 6.2 Hardcoded install paths (inputs for the build-time rewrite)

- `/usr/share/opentelemetry_shell` appears in 10 runtime `.sh` files:
  - `api.sh` 4, `agent.sh` 4, python 8, node 5, java 3, curl, wget, netcat, docker, gh-aw docker.
  - Both Java sources.
  - `meta/debian/postinst` (19) and `prerm`.
  - `actions/instrument/job/{inject_and_init,decorate_action_run,decorate_action_node,decorate_action_docker}.sh`, `sdk_factory.py` (2), `actions/instrument/workflow/Dockerfile`.
  - `build.yml` (41) and `test_github.yml`.
- `/opt/opentelemetry_shell/venv` is hardcoded in `api.sh`. It is fine for the tarball install (`/opt` isn't sealed), but not for Homebrew (`#{prefix}/opt/...`).
- `/usr/bin/opentelemetry_shell*.sh` / `otel*.sh`: 63 test files.
- Docker/gh-aw mount targets must stay Linux paths regardless (#3947).

### 6.3 GNU userland dependency (decision needed)

- The runtime relies on GNU behavior in many places:
  - `mktemp -u -p` (31× across `api.sh`, `api.observe.*`, curl)
  - `date +%s.%N` (curl, wget, gh)
  - `readlink -f`, `md5sum`, `find -executable`
  - `sed -i` (github.actions.* instrumentations), `tac`, `seq`
  - `env --ignore-signal` (feature-detected already)
  - `xargs -0` on `/proc` data
- The approach both lines chose is to require Homebrew GNU tools with gnubin first on `PATH`. Consequences to decide or document:
  - Users must export `PATH` themselves (caveats/README), or `otel.sh` could prepend gnubin itself (defined once at load).
  - `sudo` may reset `PATH` (affects postinst, §5.3).
  - The formula doesn't declare every optional tool the runtime probes for (e.g. `flock`; `xxd` is only `:recommended`), and `timeout` only exists as `gtimeout` unless gnubin is on `PATH`.
- Tools absent on macOS, where the existing guards already handle it:
  - `strace` (subprocess/signal observation is simply unavailable; DTrace/dtruss need SIP disabled, so not an option).
  - `ldd`, `dpkg`, `rpm`, `apk`.
  - `flock`, `ionice`, `taskset`, `busybox`, `podman`, `dumb-init`, `parallel` — `type` checks only. @smith's log showed unredirected `type` noise for these in 2024.

### 6.4 Release and distribution

- Add the tarball, and possibly the templated formula, to `softprops/action-gh-release` `files:`.
- Add `attest-build-provenance` for `*.tar.gz`.
- Decide the tap repo (#3952) and its publish step.
- Handle the Apple Silicon prefix under Homebrew.
- Remove the `brew tap … Thoth` URL from INSTALL.sh, the action installer and the README until the tap exists.

### 6.5 Shells and environments on macOS

- `/bin/bash` is bash **3.2** and SIP-protected. `/bin/sh` is bash 3.2 in POSIX mode by default. The default login shell is **zsh**.
- CI only tests Homebrew bash (`shell: /opt/homebrew/bin/bash`), and the matrix only runs `bash`.
- Decide which shells are supported on macOS (Homebrew bash only? system bash 3.2? zsh?) and document it. Scripts with `#!/bin/bash` or `#!/bin/sh` shebangs will run under bash 3.2 and lose `DYLD_*` (D13).
- No Docker on GitHub-hosted macOS runners: docker tests self-skip (`type docker`). The docker/gh-aw asymmetric mount change (#3947) is therefore untestable in CI.
- Linux-only tests must be skipped explicitly: debconf (self-skips on a missing `confmodule`), dash on non-Debian (`run_tests.sh`), anything using `/proc`, `apt-file`, `strace`, `taskset`/`ionice`, etc.

### 6.6 GitHub Actions instrumentation on macOS runners

This is out of scope of the shell port but touched by #3174. See D19. The job and workflow actions are no-ops on macOS today, and the `job-macos` tests from #3550/#3894 only assert that. A real port also needs:

- otelcol-contrib for darwin
- no apt cache
- `/proc`-based Runner.Worker detection replaced
- the `forward.c` build for darwin (not investigated)

---

## 7. Suggested restart order (a suggestion, not decided)

1. Decide the layout: tarball → `/usr/local/...` + `/opt/opentelemetry_shell`, with the maintainer's build-time path rewrite (e.g. `sed` in `build-brew` over §6.2, plus the baked-in version). Decide the GNU-tools policy (§6.3) and the supported shells (§6.5).
2. Build: `build-http-macos` (consider a universal dylib), a per-file `merge-http`, `.so`-less arch skip in deb/rpm, `build-brew` with java, node and python artifacts and the top-level `postinst`, tarball members without the `./` ambiguity. Publish and attest the tarball in releases.
3. Runtime: "define once" `[ -d /proc ]` / Darwin function variants for every item in §6.1. Fix the BSD `find` executable test. Rework HTTP injection for macOS or explicitly disable it there (D13).
4. Install: `INSTALL.sh` tarball path plus the real `postinst` (prefix-aware, `cp -u` portability). Keep the CI test install step identical to what users run.
5. Tests: entrypoints resolvable on mac, Linux-only tests skipped, only `macos-latest` (plus a pinned `macos-15`/`macos-26` if wanted), and CI actually running on the PR.
6. Homebrew tap (#3952) as a follow-up consuming the released tarball.
7. Separately: the GitHub Actions instrumentation on macOS runners (D19, §6.6).

(🤖 generated by Claude)

**Comment by plengauer** [comment]
## Corrections and clarifications to the consolidated summary

Re-verified the [consolidated summary](https://github.com/plengauer/Thoth/issues/550#issuecomment-5634336773) against the repo and the PRs again (2026-09-11 ~22:30Z). Nothing changed on any of the 11 PRs in the meantime — all 11 head SHAs are still the ones listed there, and no new comments, reviews or CI runs appeared. The `/proc` line numbers in §6.1 and the counts in §6.2 still resolve on today's `main` (`b2e4026`); the only changes to `main` since `e7dd341` are workflow files (`#4202`, `#4215`), which do not touch any file the summary cites. Five things need correcting or sharpening.

**1. §5.1 — "Copilot, 2026-06-16 (18 comments), all fixed" is too strong.** Of those 18 review threads:

- 15 were fixed in code (`84f396fa`/`2a1b9fb1`/`3220a0c1`),
- 3 were resolved as noise rather than fixed — the `ldd`-on-Darwin comment on `agent.instrumentation.curl.sh`, the identical one on `agent.instrumentation.wget.sh`, and the unredirected `type dpkg` in `actions/instrument/workflow/main.sh` (these are the ones listed at the bottom of §5.2),
- and 4 threads are still open in the UI even though the code fix landed: `clang -shared`, the `merge-http` nesting, the `run_tests.sh` timeout fallback, and the README gnubin note.

**2. §6.2 — those numbers are matching *lines*, not occurrences.** For a find-and-replace checklist the occurrence counts are the ones that matter, and several are higher:

| File | lines | occurrences of `/usr/share/opentelemetry_shell` |
|---|---|---|
| `agent.instrumentation.java.sh` | 3 | **10** |
| `meta/debian/postinst` | 19 | **26** (this is where #3948's "26 literals" figure comes from) |
| `.github/workflows/build.yml` | 41 | **44** |
| `meta/debian/prerm` | 1 | **3** |
| `agent.sh` | 4 | **5** |
| `agent.instrumentation.docker.sh` | 1 | **2** |
| `agent.instrumentation.github.aw.docker.sh` | 1 | **2** |
| `api.sh` 4, python 8, node 5, curl 1, wget 1, netcat 1 | | unchanged |

Likewise the test entrypoints: 63 files under `tests/`, but **65** occurrences of `/usr/bin/opentelemetry_shell*.sh` / `/usr/bin/otel*.sh`.

**3. §3.1 — "#3174 swapped `find -printf` for `ls | grep '^http_'`" mischaracterises it.** `ls | grep '^http_'` is the *older* form that `main` itself used back in March; `main` has used `find . -maxdepth 1 -name 'http_*' -type d -printf '%f\n'` since then. So the `ls | grep` in #3174's head is a revert introduced by the August rebase, not a deliberate choice — keep `main`'s `find -printf` form on the restart. (The snippet quoted there also drops a trailing `|| true`: the actual line is `rmdir "$directory" || true`.)

**4. §2 — "Net diff vs `main` is 16 files, +339/−56" is measured against the merge base**, `a765a510` (2026-08-26), not against `main`'s tip. The figure itself is correct. Side note in the same section: #4202 has since been merged, which confirms the note that the `experiment` failure raised on #3174 was unrelated to macOS.

**5. Addition to §6.5 — netcat.** `agent.instrumentation.netcat.sh` aliases `nc`, `ncat` and `netcat`, and its option table mixes BSD-`nc`, GNU-`netcat` and `ncat` spellings (e.g. `-i` vs `--idle-timeout`). macOS ships BSD `nc` only, so on a Mac the alias that will actually fire is the one whose flag parsing was least exercised; `ncat`/`netcat` come from Homebrew. The same file also uses `timeout`, `xxd`, `dd` and `head -c` (see §6.3). Worth an explicit check on a runner — this was not raised in any PR or review, and no test has ever run it on macOS.

Everything else in the summary re-verified as written, including the six `Authorization: ******` occurrences on four lines (D10), the `publish.yml` asset list (D8 — still `*.deb`, `*.rpm`, `*.apk` with `fail_on_unmatched_files` on today's `main`), the `/proc` inventory (§6.1), the `mktemp -u -p` count of 31 (§6.3), the postinst literal count at `d0d4253`, and #3559 being an automated backport of #3550 to `release/v5.55`.

(🤖 generated by Claude)
